server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java

```diff
@@ -748,6 +819,151 @@ private Backup.VolumeInfo getVolumeInfo(List<Backup.VolumeInfo> backedUpVolumes,
+            if (diskOffering == null) {
+                throw new CloudRuntimeException("Unable to find the disk offering with uuid (" + diskOfferingIds[i] + ") stored in backup. Please specify a valid disk offering id while creating the instance");
+            }
+            Long size = Long.parseLong(diskSizes[i]) / (1024 * 1024 * 1024);
+            Long minIops = (diskOffering.isCustomizedIops() != null && diskOffering.isCustomizedIops() && !minIopsList[i].equals("null")) ?
+                    Long.parseLong(minIopsList[i]) : null;
+            Long maxIops = (diskOffering.isCustomizedIops() != null && diskOffering.isCustomizedIops() && !maxIopsList[i].equals("null")) ?
+                    Long.parseLong(maxIopsList[i]) : null;
+            diskOfferingInfoList.add(new DiskOfferingInfo(diskOffering, size, minIops, maxIops, deviceId));
+        }
+        return diskOfferingInfoList;
+    }
+
+    @Override
+    public boolean restoreBackupToVM(final Long backupId, final Long vmId) throws ResourceUnavailableException {
+        final BackupVO backup = backupDao.findById(backupId);
+        if (backup == null) {
+            throw new CloudRuntimeException("Backup " + backupId + " does not exist");
+        }
+        validateBackupForZone(backup.getZoneId());
+
+        VMInstanceVO vm = vmInstanceDao.findByIdIncludingRemoved(vmId);
+        if (vm == null) {
+            throw new CloudRuntimeException("VM ID " + backup.getVmId() + " couldn't be found on existing or removed VMs");
+        }
+        accountManager.checkAccess(CallContext.current().getCallingAccount(), null, true, vm);
+
+        if (vm.getRemoved() != null) {
+            throw new CloudRuntimeException("VM is removed from CloudStack");
+        }
+        if (vm.getRemoved() != null) {
+            throw new CloudRuntimeException("VM should be in the stopped state");
+        }
+        if (!vm.getState().equals(VirtualMachine.State.Stopped)) {
+            throw new CloudRuntimeException("Existing VM should be stopped before being restored from backup");
+        }
+
+        List<Backup.VolumeInfo> backupVolumes = backup.getBackedUpVolumes();
+        if (backupVolumes == null) {
+            throw new CloudRuntimeException("Backed-up volumes not found");
+        }
+
+        List<VolumeVO> vmVolumes = volumeDao.findByInstance(vmId);
+        if (vmVolumes.size() != backupVolumes.size()) {
+            throw new CloudRuntimeException("Unable to restore VM with the current backup as the backup has different number of disks as the VM");
+        }
+
+        BackupOffering offering = backupOfferingDao.findByIdIncludingRemoved(backup.getBackupOfferingId());
+        if (offering == null) {
+            String errorMessage = "Failed to find backup offering of the VM backup.";
+            throw new CloudRuntimeException("Failed to find backup offering of the VM backup.");
+        }
+        if ("networker".equals(offering.getProvider())) {
```

Comment:
maybe better to define in the provider itself - `offering.getProvider().supportsInstanceFromBackup()`